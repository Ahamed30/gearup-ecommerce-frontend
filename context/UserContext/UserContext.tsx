import {
  createUserWithEmailAndPassword,
  onAuthStateChanged,
  signInWithEmailAndPassword,
  signOut,
  updateProfile,
} from "firebase/auth";
import { useRouter } from "next/navigation";
import { createContext, useContext, useEffect, useState } from "react";
import type { ReactNode } from "react";
import { auth } from "@/firebase/firebaseConfig";
import { AddressType, UserType } from "@/types";
import { useApp } from "../AppContext";

interface UserContextType {
  user: UserType | null;
  handleChangeData: (e: {
    target: { name: string; value: string; type: string; checked: boolean };
  }) => void; // when updating via backend these can be update to Promise<void> from void
  updateAddress: (address: AddressType, typeOfAddress: string) => void;
  isLoggedIn: boolean;
  handleLogin: (email: string, password: string) => void;
  handleSignUp: (
    name: string,
    email: string,
    password: string
  ) => Promise<boolean>;
  handleSignOut: () => void;
  setIsLoggedIn: (loginStatus: boolean) => void;
}

//@ts-expect-error: ignore initial context creation
const UserContext = createContext<UserContextType>(null);

export const useUser = () => useContext(UserContext);

export interface UserContextProviderProps {
  children: ReactNode;
}

export const UserContextProvider = ({ children }: UserContextProviderProps) => {
  const [user, setUser] = useState<UserType>({
    email: "",
    deliveryAddress: null,
    phoneNumber: 1234567890,
    billingAddress: null,
    sameBillingDelivery: true,
    receiveNewsletter: false,
  });

  const [isLoggedIn, setIsLoggedIn] = useState<boolean>(false);
  const router = useRouter();
  const { setIsLoading } = useApp();

  useEffect(() => {
    onAuthStateChanged(auth, (user) => {
      if (user) {
        setUser((prevData) => {
          return { ...prevData, email: String(user.email) };
        });
        setIsLoggedIn(true);
      } else {
        setUser((prevData) => {
          return { ...prevData, email: "" };
        });
      }
    });
  }, []);

  const handleChangeData = (e: {
    target: { name: string; value: string; type: string; checked: boolean };
  }) => {
    const { name, value, type, checked } = e.target;
    setUser((prevData) => ({
      ...prevData,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const updateAddress = (address: AddressType, typeOfAddress: string) => {
    setUser((prevCart) => {
      const updatedUser: UserType = { ...prevCart };

      if (typeOfAddress === "deliveryAddress") {
        updatedUser.deliveryAddress = address;
      } else {
        updatedUser.billingAddress = address;
      }

      return updatedUser;
    });
  };

  const handleLogin = async (email: string, password: string) => {
    setIsLoading(true);
    await signInWithEmailAndPassword(auth, email, password)
      .then(async (userCredential) => {
        fetch("/api/login", {
          method: "POST",
          headers: {
            Authorization: `Bearer ${await userCredential.user.getIdToken()}`,
          },
        }).then((response) => {
          if (response.status == 200) {
            setIsLoggedIn(true);
            return response;
          }
        });
      })
      .catch((error) => console.error(error));
    window.location.replace(sessionStorage.getItem("previousPage") ?? "/");
    setIsLoading(false);
  };

  const handleSignUp = async (
    name: string,
    email: string,
    password: string
  ) => {
    setIsLoading(true);
    const userCredential = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    )
      .then(async (userCredential) => {
        handleLogin(email, password);
        return userCredential;
      })
      .catch((error) => {
        const errorCode = error.code;
        const errorMessage = error.message;
        if (
          errorCode === "auth/email-already-exists" ||
          errorCode === "auth/email-already-in-use"
        ) {
          return false;
        }
        console.error("Unable to register", errorMessage);
      });

    if (userCredential) {
      await updateProfile(userCredential.user, {
        displayName: name,
      });
    }
    router.push(sessionStorage.getItem("previousPage") ?? "/");
    setIsLoading(false);
    return true;
  };

  const handleSignOut = async () => {
    setIsLoading(true);
    await signOut(auth);

    await fetch("/api/logout", {
      method: "POST",
    }).catch((err) => {
      console.error("Failed to SignOut", err);
    });
    window.location.reload();
    setIsLoggedIn(false);
    setIsLoading(false);
  };

  return (
    <UserContext.Provider
      value={{
        user,
        handleChangeData,
        updateAddress,
        isLoggedIn,
        setIsLoggedIn,
        handleLogin,
        handleSignUp,
        handleSignOut,
      }}
    >
      {children}
    </UserContext.Provider>
  );
};
