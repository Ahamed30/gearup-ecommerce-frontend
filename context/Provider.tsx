"use client";

import { SessionProvider } from "next-auth/react";
import type { ReactNode } from "react";
import { AppContextProvider } from "./AppContext";
import { CartContextProvider } from "./CartContext";
import { UserContextProvider } from "./UserContext";

interface ProviderProps {
  children: ReactNode;
}

export const Provider = ({ children }: ProviderProps) => {
  return (
    <SessionProvider>
      <AppContextProvider>
        <UserContextProvider>
          <CartContextProvider>{children}</CartContextProvider>
        </UserContextProvider>
      </AppContextProvider>
    </SessionProvider>
  );
};
