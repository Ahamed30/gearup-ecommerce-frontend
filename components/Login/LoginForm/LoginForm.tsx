"use client";

import Image from "next/image";
import { Button } from "@/components/Button";
import { CustomCheckbox } from "@/components/CustomCheckBox";
import { TextInput } from "@/components/TextInput";
import { Typography } from "@/components/Typography";
import { useUser } from "@/context/UserContext";
import { FormEvent, useEffect, useState } from "react";
import { useRouter } from "next/navigation";

export const LoginForm = () => {
  const { isLoggedIn, setIsLoggedIn } = useUser();
  const router = useRouter();

  const AdditionalLogins = () => {
    const additionalLoginIcons = ["google", "apple", "facebook"];
    return additionalLoginIcons.map((loginIcon, index) => {
      return (
        <div
          key={index}
          className="border border-[#232321] rounded-xl py-4 px-10 lg:px-12 cursor-pointer"
        >
          <Image
            src={`/${loginIcon}-login.svg`}
            height={32}
            width={32}
            alt={`${loginIcon} Login`}
          />
        </div>
      );
    });
  };

  const handleSubmit = (e: FormEvent) => {
    // TODO: handle login operations
    // if valid
    e.preventDefault();
    setIsLoggedIn(true);
    router.push("/");
  };

  useEffect(() => {
    if (isLoggedIn) {
      router.push("/");
    }
  }, [isLoggedIn]);

  return (
    <div className="w-full lg:w-1/2 px-4 lg:px-10">
      <Typography
        variant="headline"
        className="text-2xl md:text-3xl font-bold mb-4"
      >
        Login
      </Typography>
      <Typography
        variant="headline"
        className="text-base md:text-xl font-semibold underline cursor underline-offset-5 cursor-pointer"
      >
        Forgot your password
      </Typography>
      <div className="flex flex-col gap-6 py-6">
        <form className="flex flex-col gap-6" onSubmit={(e) => handleSubmit(e)}>
          <TextInput placeholder="Email" type="email" required />
          <TextInput
            className="w-full lg:w-5/6"
            placeholder="Password"
            type="password"
            required
          />
          <CustomCheckbox text="Keep me logged in - applies to all log in options below. More info" />
          <Button
            color="primary"
            type="submit"
            className="w-full lg:w-5/6 flex justify-between items-center cursor-pointer"
          >
            <Typography
              color="#FFFFFF"
              variant="headline"
              className="text-sm lg:text-base uppercase"
            >
              Email Login
            </Typography>
            <Image
              src="/arrow-forward.svg"
              height={16}
              width={16}
              alt="Go forward"
            />
          </Button>
        </form>
        {/* {TODO: Need to change the UI} */}
        <div className="flex justify-between gap-x-2">
          <AdditionalLogins />
        </div>
        <Typography>
          {`By clicking 'Log In' you agree to our website KicksClub Terms &
          Conditions, Kicks Privacy Notice and Terms & Conditions.`}
        </Typography>
      </div>
    </div>
  );
};
