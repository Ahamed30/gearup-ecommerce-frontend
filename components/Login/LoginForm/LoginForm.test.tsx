import { render } from "@testing-library/react";
import { LoginForm } from "./LoginForm";
import { UserContextProvider } from "@/context/UserContext";

const mockRouterPush = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

const renderComponent = () => {
  return render(
    <UserContextProvider>
      <LoginForm />
    </UserContextProvider>
  );
};

describe("Login Form", () => {
  it("should match the snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
