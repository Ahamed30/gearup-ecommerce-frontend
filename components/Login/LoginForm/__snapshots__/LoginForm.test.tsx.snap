// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Form should match the snapshot 1`] = `
<div>
  <div
    class="w-full lg:w-1/2 px-4 lg:px-10"
  >
    <p
      class="className text-2xl md:text-3xl font-bold mb-4"
      style="color: rgb(35, 35, 33);"
    >
      Login
    </p>
    <p
      class="className text-base md:text-xl font-semibold underline cursor underline-offset-5 cursor-pointer"
      style="color: rgb(35, 35, 33);"
    >
      Forgot your password
    </p>
    <div
      class="flex flex-col gap-6 py-6"
    >
      <form
        class="flex flex-col gap-6"
      >
        <div
          class="flex flex-col"
        >
          <div
            class="px-2 py-2.5 rounded-lg border justify-start items-center inline-flex border-[#232321]"
          >
            <input
              class="className w-full px-2 placeholder-[#79767C] text-[16px] font-normal tracking-wide bg-transparent focus:outline-none"
              placeholder="Email"
              required=""
              style="color: rgb(31, 26, 36);"
              type="email"
            />
          </div>
        </div>
        <div
          class="flex flex-col w-full lg:w-5/6"
        >
          <div
            class="px-2 py-2.5 rounded-lg border justify-start items-center inline-flex border-[#232321]"
          >
            <input
              class="className w-full px-2 placeholder-[#79767C] text-[16px] font-normal tracking-wide bg-transparent focus:outline-none"
              placeholder="Password"
              required=""
              style="color: rgb(31, 26, 36);"
              type="password"
            />
          </div>
        </div>
        <label
          class="flex items-center cursor-pointer"
        >
          <div
            class="relative"
          >
            <input
              class="hidden"
              type="checkbox"
            />
            <img
              alt="Checked this box"
              data-nimg="1"
              decoding="async"
              height="24"
              loading="lazy"
              src="/checkbox-false.svg"
              style="color: transparent;"
              width="24"
            />
          </div>
          <span
            class="className ml-2 className text-base font-semibold"
            style="color: rgb(35, 35, 33);"
          >
            Keep me logged in - applies to all log in options below. More info
          </span>
        </label>
        <button
          class="className text-white rounded-[8px] px-3 lg:px-6 bg-[#232321] hover:bg-[#4A69E2] text-xs lg:text-sm py-[12px] w-full lg:w-5/6 flex justify-between items-center cursor-pointer"
          type="submit"
        >
          <p
            class="className text-sm lg:text-base uppercase"
            style="color: rgb(255, 255, 255);"
          >
            Email Login
          </p>
          <img
            alt="Go forward"
            data-nimg="1"
            decoding="async"
            height="16"
            loading="lazy"
            src="/arrow-forward.svg"
            style="color: transparent;"
            width="16"
          />
        </button>
      </form>
      <div
        class="flex justify-between gap-x-2"
      >
        <div
          class="border border-[#232321] rounded-xl py-4 px-10 lg:px-12 cursor-pointer"
        >
          <img
            alt="google Login"
            data-nimg="1"
            decoding="async"
            height="32"
            loading="lazy"
            src="/google-login.svg"
            style="color: transparent;"
            width="32"
          />
        </div>
        <div
          class="border border-[#232321] rounded-xl py-4 px-10 lg:px-12 cursor-pointer"
        >
          <img
            alt="apple Login"
            data-nimg="1"
            decoding="async"
            height="32"
            loading="lazy"
            src="/apple-login.svg"
            style="color: transparent;"
            width="32"
          />
        </div>
        <div
          class="border border-[#232321] rounded-xl py-4 px-10 lg:px-12 cursor-pointer"
        >
          <img
            alt="facebook Login"
            data-nimg="1"
            decoding="async"
            height="32"
            loading="lazy"
            src="/facebook-login.svg"
            style="color: transparent;"
            width="32"
          />
        </div>
      </div>
      <p
        class="className"
        style="color: rgb(35, 35, 33);"
      >
        By clicking 'Log In' you agree to our website KicksClub Terms &
          Conditions, Kicks Privacy Notice and Terms & Conditions.
      </p>
    </div>
  </div>
</div>
`;
